Static analysis result bug fixes

PavolSloboda opened the pull request:
I have received a run of static analysis using the openscanhub tool:
https://openscanhub.fedoraproject.org/task/49253/
which i have evaluated and fixed as many errors and warnings as I could.
Mostly minor tweaks about memory handling and allocation.

PavolSloboda committed:
fix: fixed the possible nullPointerAritheticOutOfMemory by using
SET_CLIENT_ERROR and return 1;
because jumping to the error label is not viable in this case as pos is
not initialized and it could raise more errors, this way we still handle
the error while not clashing with the original error handling
implementation

PavolSloboda committed:
fix: fixed the va_end_missing on line 4626 by adding a va_end call

PavolSloboda committed:
fix: added a failsafe return 1; in case the malloc at 559 fails so that
we don't work with NULL

PavolSloboda committed:
fix: fixed the posibility of undefined behaviour of the strchr previously
at 252 if the strdup at 248 fails and thus returns NULL by
setting a client error signalling that the client ran out of
memory for easier communication of what exactly failed during the
loading of the plugins and why. Returning from the function just after

PavolSloboda committed:
fix: fixed a memory leak due to the socket not being closed after
encountering an error by closing the socket before return

PavolSloboda committed:
fix: fixed a possible leak when one of the files (fp1 or fp2) fails to
open by adding a simple condition that checks whether one of them opens
while the other fails and closing it
and added the extra prentheses for the assignment inside the if statement
for easier debugging and to avoid static analysis flagging it as a wrong
statement inside the if clause

PavolSloboda committed:
fix: added the closing of the test_file before various fail statements
to avoid memory leaks

PavolSloboda committed:
fix: added frees for any allocated memory if the allocation of other
memory fails and the FAIL_UNLESS calls needed the stop the test if this
happens

PavolSloboda committed:
fix: added the closing of the file descriptor if the check_mysql_rc
macro will return FAIL to avoid memory leaks

PavolSloboda committed:
fix: added the necessary frees for the bind var before any of the
check_stmt_rc() and check_mysql_rc() macros potentialy return from the
fuction

PavolSloboda committed:
fix: fixed the posible read of NULL when strdup fails thus resulting in
rpl->filename being NULL by setting an error signalling that we ran out
of memory to ease the debugging of the mariadb_rpl_options function and make
it clearer if it ever fails because of this. Also added va_end in this case to
avoid the va_list (ap) leaking when we return from this function and
then returning 1 from the fucntion.

PavolSloboda committed:
fix: fixed the order of the mariadb_free_rpl_event() and rpl_set_error()
calls to avoid using the rpl_even var after freeing it

grooverdan commented on unittest/libmariadb/misc.c:
```diff
@@ -287,21 +287,31 @@ static int test_frm_bug(MYSQL *mysql)
 
   row= mysql_fetch_row(result);
+  if(!row)
+    fclose(test_file);
```
there would probably be better as goto statements towards end a with an error string pointer set.  The `mysql_free_result` is needed  and the `mysql_stmt_close` to cleanup from the `stmt`s.

PavolSloboda replied:
I'm sorry I don't think I quite understand what you mean. The check_mysql_rc(), FAIL_IF() adn FAIL_UNLESS() functions seem to be the ones handling the exits. I just added cleanup for the opened test_file.

grooverdan replied:
While you've address the `fclose` aspects of this test case the cleanup is more substantial. `unlink(test_frm)` is part of the cleanup.

After `mysql_store_result` is called, `mysql_free_result(result);` is part of the cleanup. noting `row` is a memory pointer within `result`.

The last `mysql_stmt_close(stmt);`  could be moved to after the last `mysql_stmt_fetch` (and remove from `fopen` failure cleanup) to avoid making this part of the general cleanup.

So completing the cleanup suggestion:

at beginning:
```
int fn_ret=OK;
result =NULL;
```


at end:
```
error:
if (result) mysql_free_result(result);
fclose(test_file);
mysql_query(mysql, "drop table if exists test_frm_bug");
unlink(test_frm);
return fn_ret;  
```

And `FAIL_UNLESS`, `FAIL_IF`, `check_mysql_rc` are replaced with similar macros, except for `return FAIL` and becomes, `fn_ret= FAIL; goto error` for the usages after the file is open.

grooverdan reviewed (commented):
rest looks quite good on my look though. Thank you.

grooverdan commented on libmariadb/mariadb_lib.c:
```diff
@@ -886,6 +886,8 @@ static int parse_connection_string(MYSQL *mysql, const char *unused __attribute_
   /* don't modify original dsn */
   conn_save= (char *)malloc(len + 1);
+  if (!conn_save)
+    goto error;
```
by adding this `goto`, the first line at the `error:` label requires that `pos` is initialized, which it isn't at this code location.

suggest `SET_CLIENT_ERROR(mysql, CR_OUT_OF_MEMORY, SQLSTATE_UNKNOWN, 0);` and then jump to a new label past error or just `return 1`

PavolSloboda replied:
Thank you, I have fixed this in the newest rebase the way you suggested. I will continue to address the rest of these comments.

grooverdan commented on unittest/libmariadb/connection.c:
```diff
@@ -1540,7 +1541,19 @@ static int test_conc327(MYSQL *unused __attribute__((unused)))
   fp1= fopen(cnf_file1, "w");
   fp2= fopen(cnf_file2, "w");
-  FAIL_IF(!fp1 || !fp2, "fopen failed");
+  if(failed_opening_files = !fp1 || !fp2)
```
The usual practice for doing assignments within a `if` is to add additional parentheses to be explict about this not being a accidental == typo.
`if((failed_opening_files = !fp1 || !fp2))`

PavolSloboda replied:
Thanks for the heads up, I have only learned recently about this practice, fixed it with the newest rebase. I have also squashed this and the next commit together as they resolve the same problem and fixed the parentheses in it as well. I will continue to address the rest of these comments.

grooverdan commented on unittest/libmariadb/connection.c:
```diff
@@ -1565,7 +1578,18 @@ static int test_conc327(MYSQL *unused __attribute__((unused)))
   fp1= fopen(cnf_file1, "w");
   fp2= fopen(cnf_file2, "w");
-  FAIL_IF(!fp1 || !fp2, "fopen failed");
+  if(failed_opening_files = !fp1 || !fp2)
```
as above - extra parentheses, or FAIL_IF into each branch and omit `failed_opening_files`.

grooverdan commented on libmariadb/ma_client_plugin.c.in:
```diff
@@ -248,6 +248,9 @@ static void load_env_plugins(MYSQL *mysql)
   free_env= strdup(s);
   plugs= s= free_env;
 
+  if (!free_env)
```
since we have a `mysql` pointer, we can push an error here `SET_CLIENT_ERROR(mysql, CR_OUT_OF_MEMORY, SQLSTATE_UNKNOWN, 0);`

PavolSloboda replied:
Thanks, I added this fix in the newest rebase, I will continue looking into the other comments.

grooverdan commented on libmariadb/mariadb_rpl.c:
```diff
@@ -1980,6 +1980,8 @@ int STDCALL mariadb_rpl_optionsv(MARIADB_RPL *rpl,
     else if (arg1)
     {
       rpl->filename= strdup((const char *)arg1);
+      if (!rpl->filename)
```
set `rpl_set_error(rpl, CR_OUT_OF_MEMORY, 0);` before return

PavolSloboda replied:
Thank you, I have added this in the latest rebase. I have also added the closing of the ap va_list before the return to avoid any possible memory leaks. I will continue looking into the other comments.

grooverdan commented on unittest/libmariadb/ps_bugs.c:
```diff
@@ -2681,7 +2682,17 @@ static int test_bug5194(MYSQL *mysql)
   param_str= (char*) malloc(COLUMN_COUNT * CHARS_PER_PARAM);
 
-  FAIL_IF(my_bind == 0 || query == 0 || param_str == 0, "Not enough memory");
+  if(allocation_failed = !my_bind || !query || !param_str)
```
extra parentheses or:

```
FAIL_UNLESS(my_bind, "Not enough memory");
query = ...
if (!query)
  free(my_bind)
FAIL_UNLESS(query, "Not enough memory");
param_str = ...
if (!param_str)
{
  free(my_bind)
  free(query)
}
FAIL_UNLESS(param_str, "Not enough memory");
```

PavolSloboda replied:
Thank you. I have updated this in the latest rebase. I will continue looking into the other comments.

grooverdan commented on unittest/libmariadb/ps_bugs.c:
```diff
@@ -5166,16 +5177,22 @@ static int test_maxparam(MYSQL *mysql)
   bind = calloc(65535, sizeof *bind);
```
how did this escape static analysis?

`FAIL_UNLESS(bind, "Not enough memory")`

PavolSloboda replied:
Thanks. This one might be on me as I have misinterpreted the output of the static analysis which was pointing to the expansion of the macro ```check_mysql_rc```. But looking at it now it was talking about this error here. I have fixed it in a new commit. I will continue looking into the other comments.

grooverdan commented on unittest/libmariadb/ps_bugs.c:
```diff
@@ -5166,16 +5177,22 @@ static int test_maxparam(MYSQL *mysql)
+    free(bind);
   check_mysql_rc(rc, mysql);
 
   buffer= calloc(1, mem);
```
and this calloc too! (others?)

PavolSloboda replied:
Thank you, this calloc was actually missed by the static analysis. Sometimes the tool misses a couple of errors. Thank you for pointing it out. I have fixed it in the latest commit. I will continue to look into the rest of the comments.

grooverdan commented on unittest/libmariadb/ps_bugs.c:
```diff
@@ -5166,16 +5177,22 @@ static int test_maxparam(MYSQL *mysql)
 
   buffer= calloc(1, mem);
   strcpy(buffer, query);
   for (i=0; i < 65534.; i++)
```
not yours, but I assume the `.` here is a typo?

PavolSloboda replied:
I am unsure about this one, we will have to wait for someone that maintains this code to clear this up.

grooverdan commented on unittest/libmariadb/ps_bugs.c:
```diff
@@ -5166,16 +5177,22 @@ static int test_maxparam(MYSQL *mysql)
   check_mysql_rc(rc, mysql);
 
   rc= mysql_query(mysql, "CREATE TABLE t1 (a int)");
+  if (rc)
```
Slightly elaborate alternate, so I'd wait for @9E0R9 to confirm as maintainer before implementing.

Change `check_mysql_rc` and `check_stmt_rc` in `unittest/libmariadb/my_test.h`  to take `...` representing pointers to be freed.
```
#define check_stmt_rc(rc, stmt, ...) \
```

Then:
`check_mysql_rc(rc, mysql, bind)` becomes the call, and `check_mysql_rc(rc, mysql, bind, buffer)` further down.

Put a function somewhere in scope
```
void free_all(void *first, ...) {
    va_list args;
    void *ptr = first;

    va_start(args, first);
    while (ptr != NULL) {
        free(ptr);
        ptr = va_arg(args, void*);
    }
    va_end(args);
}
```

the `check_mysql_rc` and `check_stmt_rc` macro will have:
`free_all(__VA_ARGS__, NULL)`
in its error path.

for correctness `check_stmt_rc` also requires `mysql_stmt_close(stmt)` before `return(FAIL)` otherwise it will eventually trigger a static analysis error.

grooverdan replied:
It wouldn't take much for the `FAIL_IF` and `FAIL_UNLESS` to take the same form. In fact for `check_stmt_rc`:
```
#define check_stmt_rc(rc, stmt, ...) FAIL_IF(rc, mysql_stmt_error(stmt), __VA_ARGS__)
```

grooverdan reviewed (commented):
note I'm not a maintainer of this code. Appreciate your time correcting the original errors.

grooverdan commented:
note in comparison to SonarQube -https://sonarcloud.io/project/issues?impactSoftwareQualities=RELIABILITY&issueStatuses=OPEN%2CCONFIRMED&id=grooverdan_mariadb-connector-c&open=AZb6Mch5sAp2y73GquHr
(stick to Bug rather that code smell, and some are incorrect)

I think these where missed:
https://sonarcloud.io/project/issues?open=AZb6Mch5sAp2y73GquHr&id=grooverdan_mariadb-connector-c
mariadb_rpl.c - missed `alloca` failure.

PavolSloboda committed:
Added the FAIL_UNLESS call for the bind variable that is being allocated
to handle the possibility of the calloc failing and ensure we are not
working with NULL in the rest of the tests

PavolSloboda committed:
Added the call to FAIL_UNLESS to handle the possibility of the
allocation of buffer failing to ensure we don't work with NULL further
in the test, also added a free to bind in this case to avoid a memory
leak
